File: packages/examples/bluesky/python/agent.py
#!/usr/bin/env python3
"""
Bluesky Agent - A full-featured AI agent running on Bluesky

This agent uses the COMPLETE elizaOS runtime pipeline:
- Full message processing through message_service.handle_message()
- State composition with all registered providers
- Action planning and execution
- Response generation via messageHandlerTemplate
- Evaluator execution
- basicCapabilities enabled by default (REPLY, IGNORE, NONE actions)

NO shortcuts, NO bypassing the pipeline - this is canonical elizaOS.
"""

from __future__ import annotations

import asyncio
import logging
import os
import signal
import sys
from pathlib import Path

from dotenv import load_dotenv

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)

# Suppress noisy HTTP logging
logging.getLogger("httpx").setLevel(logging.WARNING)
logging.getLogger("httpcore").setLevel(logging.WARNING)


def load_environment() -> None:
    """Load environment variables from .env files."""
    # Try parent directory first, then current
    parent_env = Path(__file__).parent.parent / ".env"
    local_env = Path(__file__).parent / ".env"

    if parent_env.exists():
        load_dotenv(parent_env)
    if local_env.exists():
        load_dotenv(local_env, override=True)


def validate_environment() -> None:
    """Validate required environment variables."""
    required = ["BLUESKY_HANDLE", "BLUESKY_PASSWORD"]
    missing = [key for key in required if not os.getenv(key)]

    if missing:
        logger.error(f"Missing required environment variables: {', '.join(missing)}")
        logger.error("Copy env.example to .env and fill in your credentials.")
        sys.exit(1)

    # Check for model provider
    has_model_provider = os.getenv("OPENAI_API_KEY") or os.getenv("ANTHROPIC_API_KEY")
    if not has_model_provider:
        logger.error("No model provider configured. Set OPENAI_API_KEY or ANTHROPIC_API_KEY.")
        sys.exit(1)


async def main() -> None:
    """Main entry point for the Bluesky agent."""
    print("🦋 Starting Bluesky Agent...\n")

    load_environment()
    validate_environment()

    # Import after environment is loaded
    from elizaos import string_to_uuid
    from elizaos.runtime import AgentRuntime
    from elizaos_plugin_bluesky import BlueSkyService
    from elizaos_plugin_openai import get_openai_plugin

    from character import character
    from handlers import register_bluesky_handlers

    # Get the OpenAI plugin
    openai_plugin = get_openai_plugin()

    # Create the runtime with all required plugins
    # Note: disable_basic_capabilities is False by default (provides REPLY, IGNORE, NONE actions)
    runtime = AgentRuntime(
        character=character,
        plugins=[openai_plugin],
        # These are the defaults, explicitly shown for clarity:
        disable_basic_capabilities=False,  # Keep basic actions
        enable_extended_capabilities=False,  # Extended features
    )

    # Initialize the BlueSky service
    bluesky_service = BlueSkyService.from_env()

    # Register the BlueSky service with the runtime
    if "bluesky" not in runtime.services:
        runtime.services["bluesky"] = []
    runtime.services["bluesky"].append(bluesky_service)

    # Register Bluesky event handlers
    # These handlers process notifications through the FULL elizaOS pipeline
    register_bluesky_handlers(runtime)

    # Initialize the runtime
    print("⏳ Initializing runtime...")
    await runtime.initialize()

    # Authenticate with Bluesky
    print("🔐 Authenticating with Bluesky...")
    await bluesky_service.client.authenticate()

    # Log startup info
    print(f"\n✅ Agent '{character.name}' is now running on Bluesky!")
    print(f"   Handle: {os.getenv('BLUESKY_HANDLE')}")
    print(f"   Polling interval: {os.getenv('BLUESKY_POLL_INTERVAL', '60')}s")
    print(f"   Automated posting: {os.getenv('BLUESKY_ENABLE_POSTING', 'true') != 'false'}")
    print(f"   DM processing: {os.getenv('BLUESKY_ENABLE_DMS', 'true') != 'false'}")
    print(f"   Dry run mode: {os.getenv('BLUESKY_DRY_RUN', 'false') == 'true'}")
    print("\n   Using FULL elizaOS pipeline:")
    print("   - State composition with providers")
    print("   - shouldRespond evaluation")
    print("   - Action planning & execution")
    print("   - Evaluators")
    print("\n   Press Ctrl+C to stop.\n")

    # Set up signal handlers for graceful shutdown
    shutdown_event = asyncio.Event()

    def signal_handler(sig: int, frame: object) -> None:
        logger.info(f"Received signal {sig}, initiating shutdown...")
        shutdown_event.set()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Start polling for notifications
    poll_interval = int(os.getenv("BLUESKY_POLL_INTERVAL", "60"))

    try:
        while not shutdown_event.is_set():
            try:
                # Fetch notifications
                notifications, cursor = await bluesky_service.client.get_notifications(limit=50)

                for notification in notifications:
                    if not notification.is_read:
                        # Emit event for handlers
                        await runtime.emit_event(
                            "bluesky.mention_received",
                            {
                                "runtime": runtime,
                                "source": "bluesky",
                                "notification": notification,
                            },
                        )

                # Wait before next poll
                try:
                    await asyncio.wait_for(shutdown_event.wait(), timeout=poll_interval)
                except asyncio.TimeoutError:
                    pass  # Normal timeout, continue polling

            except Exception as e:
                logger.error(f"Error during polling: {e}")
                await asyncio.sleep(poll_interval)

    finally:
        print("\n⏳ Shutting down...")
        await bluesky_service.stop()
        await runtime.stop()
        print("👋 Goodbye!")


if __name__ == "__main__":
    asyncio.run(main())

File: packages/examples/bluesky/typescript/agent.ts
#!/usr/bin/env bun

/**
 * Bluesky Agent - A full-featured AI agent running on Bluesky
 *
 * This agent uses the COMPLETE elizaOS runtime pipeline:
 * - Full message processing through messageService.handleMessage()
 * - State composition with all registered providers
 * - Action planning and execution
 * - Response generation via messageHandlerTemplate
 * - Evaluator execution
 * - basicCapabilities enabled by default (REPLY, IGNORE, NONE actions)
 *
 * NO shortcuts, NO bypassing the pipeline - this is canonical elizaOS.
 */

import { AgentRuntime, type Plugin, stringToUuid } from "@elizaos/core";
import { config } from "dotenv";

import { character } from "./character";
import { registerBlueskyHandlers } from "./handlers";

// Load environment variables from parent directory and current directory
config({ path: "../.env" });
config();

// ============================================================================
// Environment Validation
// ============================================================================

/**
 * Validate required environment variables
 */
function validateEnvironment(): void {
  const required = ["BLUESKY_HANDLE", "BLUESKY_PASSWORD"];
  const missing = required.filter((key) => !process.env[key]);

  if (missing.length > 0) {
    console.error(
      `❌ Missing required environment variables: ${missing.join(", ")}`,
    );
    console.error("   Copy env.example to .env and fill in your credentials.");
    process.exit(1);
  }

  // Check for at least one model provider
  const hasModelProvider =
    process.env.OPENAI_API_KEY || process.env.ANTHROPIC_API_KEY;

  if (!hasModelProvider) {
    console.error(
      "❌ No model provider configured. Set OPENAI_API_KEY or ANTHROPIC_API_KEY.",
    );
    process.exit(1);
  }
}

// ============================================================================
// Main Entry Point
// ============================================================================

async function main(): Promise<void> {
  console.log("🦋 Starting Bluesky Agent...\n");

  validateEnvironment();

  // Dynamically import plugins to handle workspace build order
  // These are workspace dependencies that need to be built first
  const sqlPlugin = (await import("@elizaos/plugin-sql")).default;
  const { openaiPlugin } = await import("@elizaos/plugin-openai");
  const blueskyModuleId: string = "@elizaos/plugin-bluesky";
  const { blueSkyPlugin } = (await import(blueskyModuleId)) as {
    blueSkyPlugin: Plugin;
  };

  // Create the runtime with all required plugins
  // Note: basicCapabilities is true by default (provides REPLY, IGNORE, NONE actions)
  const runtime = new AgentRuntime({
    character,
    plugins: [
      sqlPlugin, // Database persistence (PGLite by default, or Postgres)
      openaiPlugin, // LLM provider (registers TEXT_SMALL, TEXT_LARGE models)
      blueSkyPlugin, // Bluesky client (polls notifications, handles DMs)
    ],
    // These are the defaults, explicitly shown for clarity:
    // disableBasicCapabilities: false,  // Keep basic actions (REPLY, IGNORE, NONE)
    // enableExtendedCapabilities: false, // Extended features (facts, roles, etc.)
    // actionPlanning: undefined,         // Uses ACTION_PLANNING setting or defaults to true
    // checkShouldRespond: undefined,     // Uses CHECK_SHOULD_RESPOND setting or defaults to true
  });

  // Register Bluesky-specific event handlers
  // These handlers process notifications through the FULL elizaOS pipeline
  registerBlueskyHandlers(runtime);

  // Initialize the runtime
  // This starts all services including:
  // - Database adapter (creates tables, runs migrations)
  // - Model providers (registers handlers for TEXT_SMALL, TEXT_LARGE, etc.)
  // - MessageService (the core message processing pipeline)
  // - BlueSky service (authenticates, starts polling)
  console.log("⏳ Initializing runtime...");
  await runtime.initialize();

  // Ensure the Bluesky world exists for room management
  const worldId = stringToUuid("bluesky-world");
  await runtime.ensureWorldExists({
    id: worldId,
    name: "Bluesky",
    agentId: runtime.agentId,
    messageServerId: worldId,
    metadata: {
      description: "Bluesky social network",
      extra: {
        platform: "bluesky",
      },
    },
  });

  // Log startup info
  console.log(`\n✅ Agent "${character.name}" is now running on Bluesky!`);
  console.log(`   Handle: ${process.env.BLUESKY_HANDLE}`);
  console.log(
    `   Polling interval: ${process.env.BLUESKY_POLL_INTERVAL || 60}s`,
  );
  console.log(
    `   Automated posting: ${process.env.BLUESKY_ENABLE_POSTING !== "false"}`,
  );
  console.log(
    `   DM processing: ${process.env.BLUESKY_ENABLE_DMS !== "false"}`,
  );
  console.log(`   Dry run mode: ${process.env.BLUESKY_DRY_RUN === "true"}`);
  console.log("\n   Using FULL elizaOS pipeline:");
  console.log("   - State composition with providers");
  console.log("   - shouldRespond evaluation");
  console.log("   - Action planning & execution");
  console.log("   - Evaluators");
  console.log("\n   Press Ctrl+C to stop.\n");

  // Handle graceful shutdown
  const shutdown = async (signal: string) => {
    console.log(`\n${signal} received. Shutting down gracefully...`);
    try {
      await runtime.stop();
      console.log("👋 Goodbye!");
    } catch (error) {
      console.error("Error during shutdown:", error);
    }
    process.exit(0);
  };

  process.on("SIGINT", () => shutdown("SIGINT"));
  process.on("SIGTERM", () => shutdown("SIGTERM"));

  // Keep the process running
  // The BlueSky service runs polling in the background
  await new Promise(() => {});
}

main().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});
